### DISCUSSION TITLE
工具生成的代码没有参与编译

### DISCUSSION THREAD

**Discussion by Domain** [xmake-io/xmake#5670] [Q&A] [answered]
我需要用工具生成一些代码，然后再一起编译，但是现在发现生成的代码没有参与编译，需要执行两次xmake才行。
我简化的配置如下：
xmake.lua:
```lua
target("gen")
	set_kind("phony")
	set_policy("build.fence", true)
	on_build(function (target)
		os.run("gen.cmd")
	end)

target("app")
	set_kind("binary")
	add_deps("gen")
	add_files("**.cpp")
	set_policy("build.across_targets_in_parallel", false)
```

gen.cmd:
```bat
@echo off

echo void foo(); > gen.h
echo #include "gen.h" > gen.cpp
echo #include ^<stdio.h^> >> gen.cpp
echo void foo() { >> gen.cpp
echo     printf("foo\n"); >> gen.cpp
echo } >> gen.cpp
```

main.cpp:
```c++
#include <stdio.h>
#include "gen.h"

int main() {
	foo();
	return 0;
}
```
目录里面只有三个文件：xmake.lua, main.cpp, gen.cmd
第一次运行xmake，会生成gen.cpp和gen.h，但是连接报错，提示找不到foo函数，第二次运行xmake才正常。
请问这种情况需要怎么配置才正确？

**Comment by star-hengxing** [comment]
参考 xmake 源码里的 yacc/lex rule

**Comment by waruqi** [comment]
autogen 看下 https://github.com/xmake-io/xmake/tree/dev/tests/projects/other/autogen

**Reply by Domain** [reply]
是不是需要知道生成的文件名才行？我这个会生成很多文件，而且我也不知道文件名是什么

**Reply by waruqi** [reply, marked as answer]
遍历下不就好了么。。

https://github.com/xmake-io/xmake/blob/9738030c15c8e4b6d9b4a9001964cdf159155d55/xmake/rules/asn1c/xmake.lua#L45

**Reply by Domain** [reply]
谢谢，虽然实际情况比较复杂，我自己再想想办法
